Sources/Algorithms/Partition.swift

```diff
@@ -204,3 +204,171 @@ extension Collection {
+    // constant time, O(1)), we can allocate one array of size `self.count`,
+    // then insert items at the beginning or end of that contiguous block. This
+    // way, we don’t have to do any dynamic array resizing. Since we insert the
+    // right elements on the right side in reverse order, we need to reverse
+    // them back to the original order at the end.
+    
+    let count = self.count
+    
+    // The overhead of this implementation isn’t worth it for collections fewer
+    // than 8 elements. Use the simple `Sequence`-based implementation instead.
+    // This constant was determined using benchmarking. More information:
+    // https://github.com/apple/swift-algorithms/pull/152#issuecomment-887130149
+    if count < 8 {
+      return try _partitioned(belongsInSecondCollection)
+    }
+    
+    // Inside of the `initializer` closure, we set what the actual mid-point is.
+    // We will use this to partitioned the single array into two in constant time.
+    var midPoint: Int = 0
+    
+    let elements = try [Element](
+      unsafeUninitializedCapacity: count,
+      initializingWith: { buffer, initializedCount in
+        var lhs = buffer.baseAddress!
+        var rhs = lhs + buffer.count
+        do {
+          for element in self {
+            if try belongsInSecondCollection(element) {
+              rhs -= 1
+              rhs.initialize(to: element)
+            } else {
+              lhs.initialize(to: element)
+              lhs += 1
+            }
+          }
+          
+          let rhsIndex = rhs - buffer.baseAddress!
+          buffer[rhsIndex...].reverse()
+          initializedCount = buffer.count
+          
+          midPoint = rhsIndex
+        } catch {
+          let lhsCount = lhs - buffer.baseAddress!
+          let rhsCount = (buffer.baseAddress! + buffer.count) - rhs
+          buffer.baseAddress!.deinitialize(count: lhsCount)
+          rhs.deinitialize(count: rhsCount)
+          throw error
+        }
+      })
+    
+    let collections = elements.partitioned(upTo: midPoint)
+    return _tupleMap(collections, { Array($0) })
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// partitioned(upTo:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  /// Splits the receiving collection into two at the specified index
+  /// - Parameter index: The index within the receiver to split the collection
+  /// - Returns: A tuple with the first and second parts of the receiving
+  /// collection after splitting it
+  /// - Note: The first subsequence in the returned tuple does *not* include
+  /// the element at `index`. That element is in the second subsequence.
+  /// - Complexity: O(*1*)
+  @inlinable
+  public func partitioned(upTo index: Index) -> (SubSequence, SubSequence) {
+    return (
+      self[self.startIndex..<index],
+      self[index..<self.endIndex]
+    )
+  }
+}
```

Comment:
This operation is so simple that I don't think there's a point in including it here, it's not a particularly common operation and `(self[..<index], self[index...])` isn't too verbose.